test/library/draft/DistributedList/DistributedList.chpl

```diff
@@ -0,0 +1,763 @@
+            this.complete();
+
+            const numBlocks = d.size / blockSize,
+                  remainder = d.size % blockSize,
+                  numBlocksPerLoc = numBlocks / this.locDom.size;
+
+            // acquire memory to store the array
+            for (loc, idx) in zip(this.targetLocales, this.locDom) do on loc do
+                this.blockLists[idx].acquireBlocks(numBlocksPerLoc + DefaultNumBlocksPerLocale);
+
+            // store full blocks on the appropriate locales
+            for b in 0..#numBlocks {
+                const locIdx = b % this.locDom.size;
+                on this.targetLocales[locIdx] {
+                    this.blockLists[locIdx].appendBlock(arr[(b*blockSize)..#blockSize]);
+                }
+            }
+
+            // handle the final partial block if there is one
+            if remainder > 0 {
+                const rLocIdx = numBlocks % this.locDom.size;
+                on this.targetLocales[rLocIdx] {
+                    for r in 0..<remainder {
+                        const idx = blockSize * numBlocks + r,
+                              (_, blockLayerIdx, eltIdx) = indicesFor(idx);
+
+                        pragma "no auto destroy"
+                        var cpy = arr[idx];
+
+                        this.blockLists[rLocIdx].set(blockLayerIdx, eltIdx, cpy);
+                    }
+                    this.blockLists[rLocIdx].numFilled += remainder;
+                }
+            }
+
+            this.numEntries.write(d.size);
+        }
+
+        // append a new element to the end of the list
+        proc ref append(pragma "no auto destroy" in x: eltType): int {
+            this.lockAll(); defer this.unlockAll();
+            const nextIdx = this.numEntries.read(),
+                  (locIdx, blockIdx, eltIdx) = indicesFor(nextIdx);
+
+            // acquire memory if needed
+            on this.targetLocales[locIdx] do
+                if this.blockLists[locIdx].numBlocks <= blockIdx
+                    then this.blockLists[locIdx].acquireBlocks();
+
+            this.blockLists[locIdx].set(blockIdx, eltIdx, x);
+            this.blockLists[locIdx].numFilled += 1;
+
+            return this.numEntries.fetchAdd(1);
+        }
+
+        // append a group of elements to the end of the list
+        proc ref append(x: [?d] eltType): range(d.idxType)
+            where d.rank == 1
+        {
+            this.lockAll();
+
+            const nextIdx = this.numEntries.read(),
+                  newEltRange = nextIdx..#d.size,
+                  newBlocksPerLoc = (d.size / this.blockSize) / this.locDom.size + 1,
+                  (_, maxBlockIdx, _) = indicesFor(nextIdx),
+                  locRanges = rangesOnEachLocale(newEltRange, maxBlockIdx + newBlocksPerLoc);
+
+            for locIdx in this.locDom do on this.targetLocales[locIdx] {
+                // acquire memory if needed
+                if this.blockLists[locIdx].numOpenBlocks() < locRanges[locIdx].size then
```

Comment:
> Isn't this always going to evaluate to true, aside from in the case of an empty list? That seems like you're going to be resizing even if you don't necessarily need it.

Good catch, I think it should be `numBlocks` rather than `numOpenBlocks()`